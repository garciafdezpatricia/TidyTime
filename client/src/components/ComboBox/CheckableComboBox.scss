@use "../../../styles/variables" as *;

.combobox-select {
    position: relative;
    max-width: 50%;
    padding-inline: .5rem;
    border-radius: .4rem;
    border: .1rem solid $black;
    font-size: .9rem;
    background-color: $primary-grey;
}

.dropdown-header {
    display: flex;
    gap: .4rem;
    align-items: center;
    cursor: pointer;
    width: 100%;
    font-size: .8rem;

    @media screen and (min-width: 320px) and (max-width: 768px) {
        justify-content: center;
        font-size: .7rem;
    }

    @media screen and (min-width: 769px) and (max-width: 1200px) {
        font-size: .7rem;
    }
}

.dropdown-content {
    font-size: .8rem;
    position: absolute;
    top: 107%;
    right: 0;
    display: flex;
    flex-direction: column;
    width: 100%;
    background-color: $white;
    border: .1rem solid $black;
    overflow-y: auto;
    border-radius: .4rem;
    padding: .1rem .5rem;
    box-sizing: border-box;
    max-height: 7.5rem;

    @media screen and (min-width: 320px) and (max-width: 768px) {
        font-size: .7rem;
        width: fit-content;
        left: 0;
    }

    @media screen and (min-width: 769px) and (max-width: 1200px) {
        font-size: .7rem;
    }
}

.dropdown-option {
    display: flex;
    align-items: center;
    gap: 1rem;
    padding: .2rem 0rem;
    &:not(:first-child) {
        border-top: .1rem solid $black;
    }
}