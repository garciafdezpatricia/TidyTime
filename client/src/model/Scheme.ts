// Copyright 2024 Patricia García Fernández.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

export interface Task {
    id: string,
    title: string,
    desc?: string,
    labels?: Label[],
    endDate?: string,
    difficulty?: number,
    done: boolean,
    important?: boolean,
    listIndex: string,
    githubHtml?: string,
    githubUrl?: string,
    status: number;
    taskIndexInList?: number,
}

export type TaskList = {key: string, value: Task[]};

export function taskToListOfTaskList(tasks: Task[]) {
    let result:TaskList[] = [];
    for (const task of tasks) {
        // get list index from task
        const { listIndex, ...rest } = task;
        if (!result.find((list) => list.key === listIndex)) {
            result.push({key: listIndex, value: []}); // if there is not a list for the index, create it
        }
        const index = result.findIndex((list) => list.key === listIndex);
        result[index].value.push(task); // push the task
    }
    return result;
}

export interface Event {
    start: Date,
    end: Date,
    title: string,
    desc: string,
    eventId: string,
    color?: string,
    googleId?: string,
    googleHTML?: string,
    googleCalendar?: string,
    isTask?: boolean,
  }

export interface Label {
    color: string,
    name: string,
}

export interface CalendarItem { id: string, name: string, color: string};

export interface ScheduleItem {
    title: string;
    hours: string;
    minutes: string;
}