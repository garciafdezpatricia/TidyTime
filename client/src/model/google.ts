import toast from "react-hot-toast";
import { useGoogleContext } from "../components/Context/GoogleContext";
import { useEventContext } from "../components/Context/EventContext";

export function useGoogleHandler() {

    const { setCalendars, setSelectedCalendarId, setLoggedIn, setAuthUrl } = useGoogleContext();
    const { events, setEvents } = useEventContext();

    const handleLogout = () => {
        fetch('http://localhost:8080/google/auth/logout')
        .then(response => response.json())
        .then(data => {
            if (data.status === 'success'){
                localStorage.removeItem("userLoggedIn");
                // remove email param if it is in the url
                if (window.location.search.includes('email')) {
                    const newUrl = window.location.origin + window.location.pathname;
                    window.history.replaceState({}, document.title, newUrl);
                }
                setCalendars([]);
                setSelectedCalendarId("");
                toast.success("Logged out!", {
                    position: "top-center"
                })
                setLoggedIn(false);
            } else {
                toast.error("Failed to log out!", {
                    position: "top-center"
                })
            }
        }).catch(error => {
            console.error('Error when logging out');
        });
    }

    const handleLogin = () => {
        fetch('http://localhost:8080/google/auth/url')
        .then(response => response.json())
        .then(data => {
            setAuthUrl(data.authorizationUrl);
        })
        .catch(error => {
            console.error('Error al obtener la URL de autorizaciÃ³n', error);
        })
    }

    const getCalendars = (calendarId:string) => {
        return new Promise<void>((resolve, reject) => {
            fetch('http://localhost:8080/google/events/get', {
                method: 'POST',
                credentials: 'include',
                headers: { 'Content-Type': 'application/json'},
                body: JSON.stringify({calendarId}),
            })
            .then(response => response.json())
            .then(data => {
                data.map((event:any) => {
                    // the date coming from the backend is in format YYYY-MM-DDTHH:MM:SS.000Z => need to convert it to Fri Mar 29 ...
                    // when doing just new Date(...) the timezone is applied, changing the datetime fetched from the backend.
                    // The patch with the ISOString does not change datetimes.
                    event.start = new Date(new Date(event.start).toISOString().slice(0, 16));
                    event.end = new Date(new Date(event.end).toISOString().slice(0, 16));
                })
                setEvents((prev) => [...prev, ...data]);
                resolve()
            })
            .catch(error => {
                toast.error(`${error}`)
                reject(error);
            });
        })
    }

    const exportEvent = (selectedIndex:number) => {
        return new Promise<void>((resolve, reject) => {
            const eventToShare = events[selectedIndex];
            // format date to google format
            const ISOStartDate = new Date(new Date(eventToShare.start).getTime() - (new Date(eventToShare.start).getTimezoneOffset() * 60000)).toISOString();
            const ISOEndDate = new Date(new Date(eventToShare.end).getTime() - (new Date(eventToShare.end).getTimezoneOffset() * 60000)).toISOString();
            fetch('http://localhost:8080/google/events/insert', {
                method: 'POST',
                headers: {
                'Content-Type': 'application/json',
                },
                body: JSON.stringify({ 
                    start: ISOStartDate,
                    end: ISOEndDate,
                    title: eventToShare.title,
                    desc: eventToShare.desc,                     
                }),
                credentials: 'include',
            })
            .then(response => response.json())
            .then(data => {
                addGoogleInformation(data.googleId, data.googleHTML, selectedIndex);
                toast.success('Event exported to Google Calendar!', {position: "bottom-center"});
            })
            .catch(error => {
                toast.error(`${error}`);
            });
        })
    }

    /**
     * Auxiliary function to add google id and google html link to the event.
     * @param googleId contains event id on google
     * @param googleHTML contains the link to the event
     */
    const addGoogleInformation = (googleId:string, googleHTML:string, selectedIndex:number) => {
        // update event with google id and html
        let updatedEvents = [...events];
        let eventToUpdate = events[selectedIndex];
        eventToUpdate.googleId = googleId;
        eventToUpdate.googleHTML = googleHTML;
        updatedEvents = [
            ...events.slice(0, selectedIndex),
            eventToUpdate,
            ...events.slice(selectedIndex + 1)
        ];
        setEvents(updatedEvents);
    }

    return {handleLogin, handleLogout, getCalendars, exportEvent};
}