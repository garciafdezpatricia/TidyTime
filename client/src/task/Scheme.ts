export interface Task {
    title: string,
    desc?: string,
    labels?: Label[],
    endDate?: string,
    difficulty?: number,
    done: boolean,
    important?: boolean,
}

export interface Event {
    start: Date,
    end: Date,
    title: string,
    desc: string,
    eventId: string,
    color?: string,
    googleId?: string,
    googleHTML?: string,
  }

export interface Label {
    color: string,
    name: string,
}

export type TaskList = Task[];