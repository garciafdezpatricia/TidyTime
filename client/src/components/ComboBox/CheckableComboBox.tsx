import { Label } from "@/src/task/Scheme";
import { ReactNode, useContext, useState } from "react";
import { IoIosArrowDropdownCircle } from "react-icons/io";
import { useTaskContext } from "../Context/TaskContext";
import { useClickAway } from "@uidotdev/usehooks";

export interface Props{
    variant?: string,
    text: ReactNode,
    checkedLabels: Label[],
    onChange: (selectedOptions: Label[]) => void;
}

export default function CheckableComboBox({checkedLabels, onChange, text, variant} : Props) {

    const [isOpen, setIsOpen] = useState(false);
    const [selectedOptions, setSelectedOptions] = useState<Label[]>(checkedLabels);
    const {labels} = useTaskContext();
    const ref = useClickAway(() => {
        setIsOpen(false);
    })

    const toggleDropdown = () => {
        setIsOpen(!isOpen);
    };

    const handleCheckboxChange = (option:Label) => {
        const selectedIndex = selectedOptions.findIndex((label) => label.name === option.name);
        let newSelectedOptions;
        if (selectedIndex === -1) {
            newSelectedOptions = [...selectedOptions, option];  
        } else {
            newSelectedOptions = [...selectedOptions];
            newSelectedOptions.splice(selectedIndex, 1);
        }
        setSelectedOptions(newSelectedOptions);
        onChange(newSelectedOptions); // Envía la selección al componente padre
    }

    return (
        <div ref={ref} className= {variant ? variant : "combobox-select"}>
            <div className="dropdown-header" onClick={toggleDropdown}>
                {text}
            </div>
        {isOpen && (
            <div className="dropdown-content">
            {labels.map((option, index) => (
                <label className="dropdown-option" key={index}>
                <input
                    type="checkbox"
                    checked={selectedOptions.some((label) => label.name === option.name)}
                    onChange={() => handleCheckboxChange(option)}
                />
                {option.name}
                </label>
            ))}
            </div>
        )}
        </div>
    )
}