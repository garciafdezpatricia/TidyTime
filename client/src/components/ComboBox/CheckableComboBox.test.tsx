import { render } from "@testing-library/react";
import '@testing-library/jest-dom';
import CheckableComboBox from '@/src/components/ComboBox/CheckableComboBox';
import { Label } from "@/src/model/Scheme";

describe('CheckableComboBox', () => {
    it('renders without crashing', () => {
      const { container } = render(<CheckableComboBox text={undefined} checkedLabels={[]} onChange={function (selectedOptions: Label[]): void {
          throw new Error("Function not implemented.");
      } }/>);
      expect(container).toBeInTheDocument();
    });

    it('matches snapshot', () => {
        const { asFragment } = render(<CheckableComboBox text={undefined} checkedLabels={[]} onChange={function (selectedOptions: Label[]): void {
            throw new Error("Function not implemented.");
        } }/>);
        expect(asFragment()).toMatchSnapshot();
      });
  });