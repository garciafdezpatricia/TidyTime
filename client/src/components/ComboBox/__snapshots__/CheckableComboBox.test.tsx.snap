// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckableComboBox matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="combobox-select"
  >
    <div
      class="dropdown-header"
    />
  </div>
</DocumentFragment>
`;
